gen/internals.py

```diff
@@ -184,28 +212,131 @@ def add_entry(self, entry, replace_existing):
         assert isinstance(chain, list)
         self.message = message
         self.chain = chain
         super().__init__(message)
 
 
+class SkipError(Exception):
+    """Raised when this key should silently produce no errors."""
+    pass
+
+
+class Resolvable:
+    """ Keeps track of the state of the resolution of a variable with a given name
+
+    - Manages the states (unresolved vs. error / success)
+    - Keeps track of what ended up setting the value
+    - Keeps track of what caused the resolvable to enter existence (it's cause)
+    """
+
+    UNRESOLVED = 'unresolved'
+    ERROR = 'error'
+    RESOLVED = 'resolved'
+
+    def __init__(self, name, cause):
+        self._state = Resolvable.UNRESOLVED
+        self.name = name
+        self.cause = cause
+        self.error = None
+        self.setter = None
+        self._value = None
+
+    @property
+    def is_finalized(self):
+        return self._state != Resolvable.UNRESOLVED
+
+    @property
+    def is_error(self):
+        return self._state == Resolvable.ERROR
+
+    @property
+    def is_resolved(self):
+        return self._state == Resolvable.RESOLVED
+
+    def finalize_error(self, error):
+        assert self._state == Resolvable.UNRESOLVED
+        self._state = 'error'
+        self.error = error
+
+    def finalize_value(self, value: str, setter: Setter):
+        assert self._state == 'unresolved'
+        self._state = 'resolved'
+        self._value = value
+        self.setter = setter
+
+    @property
+    def value(self):
+        assert self._state == 'resolved', "is_resolved() must return true for this function to " \
+            "be called. State: {}, is_resolved: {}".format(self._state, self.is_resolved)
+
+        return self._value
+
+    def __str__(self):
+        return "<Resolvable name: {}, state: {}>".format(self.name, self._state)
+
+
+def yield_implicit_validates(target):
+    # Recursively walk the target / sub scope tree and yield a
+    # validate function for each and every switch.
+    for name, scope in target.sub_scopes.items():
+        # Note: It's important scope.cases.keys() is evaluated now, not later
+        # (is evaluated before the validate is called)
+        yield name, partial(validate_one_of, valid_values=scope.cases.keys())
+        for sub_target in scope.cases.values():
+            yield from yield_implicit_validates(sub_target)
+
+
+class ArgumentSet(dict):
+    """Manages a set of arguments and their valuables.
+
+    Makes it easy to keep track of a set of resolvables, automatically creating
+    a new one when an argument never before asked for is first accessed.
+    """
+
+    def __init__(self, get_stack):
+        self._get_stack = get_stack
+        self._finalized = False
+
+    def __missing__(self, key):
+        assert not self._finalized, "No missing keys should be accessed"
+
+        value = self[key] = Resolvable(key, self._get_stack())
+        return value
+
+    def finalize(self):
```

Comment:
The goal would be not to act like a `dict` but rather to not be able to add or remove things. Effectively become a `frozenset` of arguments. Unfortunately python doesn't allow set / frozenset to have the key not match the whole object identity...